@xen-orchestra/proxy/src/app/mixins/reverseProxy.mjs

```diff
@@ -0,0 +1,117 @@
+import { urlToHttpOptions } from 'url' // node  14.18.0, 15.7.0 is it ok ?
+import proxy from 'http2-proxy'
+
+function removeSlash(str) {
+  return str.replace(/^\/|\/$/g, '')
+}
+
+function mergeUrl(relative, base) {
+  const res = new URL(base)
+  const relativeUrl = new URL(relative, base)
+  res.pathname = relativeUrl.pathname
+  relativeUrl.searchParams.forEach((value, name) => {
+    // we do not allow to modify params already specified by config
+    if (!res.searchParams.has(name)) {
+      res.searchParams.append(name, value)
+    }
+  })
+  res.hash = relativeUrl.hash.length > 0 ? relativeUrl.hash : res.hash
+  return res
+}
+export default class ReverseProxy {
+  constructor(app, { httpServer }) {
+    this._app = app
+    httpServer.on('request', this.proxy)
+    httpServer.on('upgrade', this.upgrade)
+  }
+
+  localToBackendUrl(basePath, target, localPath) {
+    let localPathWithoutBase = removeSlash(localPath).substring(basePath.length)
+    localPathWithoutBase = './' + removeSlash(localPathWithoutBase)
+    const url = mergeUrl(localPathWithoutBase, target)
+    return url
+  }
+
+  backendToLocalPath(basePath, target, backendUrl) {
+    // keep redirect url relative to local server
+    const localPath = `${basePath}/${backendUrl.pathname.substring(target.pathname.length)}${backendUrl.search}${
+      backendUrl.hash
+    }`
+    return localPath
+  }
+
+  _getConfigFromRequest(req) {
+    for (const [path, config] of Object.entries(this._app.config.get('reverseProxies'))) {
+      const fullPath = '/proxy/v1/' + removeSlash(path)
+      if (req.url.startsWith(fullPath + '/')) {
+        if(typeof config === 'string'){
+          return {
+            path: fullPath,
+            target : config,
+            options: {}
+          }
+        }
+
+        const { target, ...options } = config
+        return {
+          path: fullPath,
+          target,
+          options,
+        }
+      }
+    }
+  }
+
+  proxy(req, res) {
+    const config = this._getConfigFromRequest(req)
+
+    if (config === undefined) {
+      return
+    }
+
+    const url = new URL(config.target)
+    const targetUrl = this.localToBackendUrl(config.path, url, req.originalUrl || req.url)
+    proxy.web(req, res, {
```

Comment:
Did you make sure that these `xo-server`'s behaviors are supported? https://github.com/vatesfr/xen-orchestra/blob/e9ca13aa12f030ae5a27b5a508373a8c20580eb2/packages/xo-server/src/index.mjs#L465-L467

`changeOrigin` is particulary important but I believe this to be the default here.

Reply:
* changeOrigin : tested, the host in the headers is ok  : `host: 'localhost:8081'`
* xfwd : also ok : `forwarded: 'by="[::ffff:127.0.0.1]:21415";for="[::ffff:127.0.0.1]:38422";proto=https;host=localhost:21415',` 


ignorePath is handled manually (especially since a request to /proxy/v1/proxy1/localPath , can be proxied to https://remotehost/base/localPath

Reply:
1. Ok
2. Ok
3. Where do you see/test that? Also, the [`xfwd`](https://github.com/http-party/node-http-proxy/#options) adds [`x-forwarded` headers](https://github.com/nxtedition/node-http2-proxy#add-x-forwarded-headers)

Reply:
with the local server in tests :
```{
  'user-agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:95.0) Gecko/20100101 Firefox/95.0',
  accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
  'accept-language': 'fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3',
  'accept-encoding': 'gzip, deflate, br',
  cookie: 'connect.sid=s%3AUlqzjRvkZRsLGTYwZ-UPT0dApjvkY47d.wv6u5q4ce2LP%2F9KWUeWK5lLPRSsIAmnoN3taiArMte0; token=E4cDGDHr0Cvr2PvTo1jLA3nhfVs4uXAm-_xkMQiMGC0; previousDisclaimer=1641919646',
  'upgrade-insecure-requests': '1',
  'sec-fetch-dest': 'document',
  'sec-fetch-mode': 'navigate',
  'sec-fetch-site': 'none',
  'sec-fetch-user': '?1',
  forwarded: 'by="[::ffff:127.0.0.1]:21415";for="[::ffff:127.0.0.1]:39284";proto=https;host=localhost:21415',
  'x-forwarded-for': '::ffff:127.0.0.1',
  'x-forwarded-proto': 'https',
  host: 'localhost:8081',
  connection: 'close'
}
```

Reply:
Can you add an explicit test for all request headers received by the server?